programs/mango-v4/src/instructions/perp_liq_base_and_positive_pnl.rs

```diff
@@ -0,0 +1,434 @@
+            .unwrap()
+            .min(liqee_base_lots.abs() - base_reduction)
+            .min(max_base_transfer.abs() - base_reduction)
+            .max(0);
+        let unweighted_change = cm!(I80F48::from(base_lots) * unweighted_health_per_lot);
+        let current_unweighted = *current_unweighted_perp_health;
+        let new_unweighted_perp = cm!(current_unweighted + unweighted_change);
+        let new_expected_perp = expected_perp_health(new_unweighted_perp);
+        let new_expected_health =
+            cm!(current_expected_health + (new_expected_perp - current_expected_perp_health));
+        msg!(
+            "{}: {} lots, health {} -> {}, unweighted perp {} -> {}",
+            step,
+            base_lots,
+            current_expected_health,
+            new_expected_health,
+            current_unweighted,
+            new_unweighted_perp
+        );
+
+        base_reduction += base_lots;
+        current_expected_health = new_expected_health;
+        *current_unweighted_perp_health = new_unweighted_perp;
+        current_expected_perp_health = new_expected_perp;
+    };
+
+    //
+    // Step 1: While the perp unsettled health is negative, any perp base position reduction
+    // directly increases it for the full amount.
+    //
+    if current_unweighted_perp_health < 0 {
+        reduce_base(
+            "negative",
+            -current_unweighted_perp_health,
+            unweighted_health_per_lot,
+            &mut current_unweighted_perp_health,
+        );
+    }
+
+    //
+    // Step 2: If perp unsettled health is positive but below max_settle, perp base position reductions
+    // benefit account health slightly less because of the settlement liquidation fee.
+    //
+    if current_unweighted_perp_health >= 0 && current_unweighted_perp_health < max_settle {
+        let settled_health_per_lot = cm!(unweighted_health_per_lot * spot_gain_per_settled);
+        reduce_base(
+            "settleable",
+            cm!(max_settle - current_unweighted_perp_health),
+            settled_health_per_lot,
+            &mut current_unweighted_perp_health,
+        );
+    }
+
+    //
+    // Step 3: Above that, perp base positions only benefit account health if the pnl asset weight is positive
+    //
+    if current_unweighted_perp_health >= max_settle && perp_market.init_pnl_asset_weight > 0 {
+        let weighted_health_per_lot =
+            cm!(unweighted_health_per_lot * perp_market.init_pnl_asset_weight);
+        reduce_base(
+            "positive",
+            I80F48::MAX,
+            weighted_health_per_lot,
+            &mut current_unweighted_perp_health,
+        );
+    }
+
+    //
+    // Execute the base reduction. This is essentially a forced trade and updates the
+    // liqee and liqors entry and break even prices.
+    //
+    let base_transfer = cm!(direction * base_reduction);
+    let quote_transfer = cm!(-I80F48::from(base_transfer) * price_per_lot * fee_factor);
```

Comment:
hmm one could aruge that you could have separate base_reductions for each step, and then have different fee_factors for them, but that would be too complicated I guess

Reply:
I think having one fee is sensible, because it mostly compensates the liqor's risk. The liqor doesn't care what step the base position came from, just that he's compensated for the risk of taking it on.